BusBoard/API/PostcodeAPIService.cs

```
@@ -0,0 +1,50 @@
+using BusBoard.Models;
+using RestSharp;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+using System.Net;
+//using System.Text.Json;
+namespace BusBoard.API;
+
+public class PostcodeAPIService
+{
+    private readonly APIService _apiService = new("https://api.postcodes.io");
+
+    public async Task<PostcodeData> GetPostcodeData(string postcode)
+    {
+        RestRequest request = new RestRequest("postcodes/{postcode}")
+            .AddUrlSegment("postcode", postcode);
+
+        RestResponse response = await _apiService.Client.GetAsync(request);
+
+        if (!response.IsSuccessful)
+        {
+            if (response.StatusCode == HttpStatusCode.NotFound)
+            {
+                throw new Exception("Error: Postcode not found");
+            }
+            
+            throw new Exception($"Request failed with status code {response.StatusCode}");
+        }
+
+        PostcodeData? data;
+
+        try
+        {
+            var parsedObject = JObject.Parse(response.Content!);
+            string resultToken = parsedObject["result"]!.ToString();
+            data = JsonConvert.DeserializeObject<PostcodeData>(resultToken);
+        }
+        catch (Exception error)
+        {
+            throw new Exception($"Error: Could not retrieve postcode data: {error.GetType} - {error.Message}");

```

Do you need to re-throw a new error?

---

As I was adding some context to describe which API service the error came from I thought this was the way to do it. Is there another way?

---

Happy to leave it :)

In a case like this you can just let an error bubble up without additional context - the callsite should tell you where the error occurred anyway